File: zenu/examples/mnist.rs
use zenu::{
    autograd::{
        activation::relu::relu, creator::from_vec::from_vec, loss::cross_entropy::cross_entropy,
        no_train, set_train, Variable,
    },
    dataset::{train_val_split, DataLoader, Dataset},
    layer::{layers::linear::Linear, Module},
    matrix::{
        device::{cpu::Cpu, nvidia::Nvidia, Device},
        num::Num,
    },
    optimizer::{sgd::SGD, Optimizer},
};
use zenu_macros::Parameters;

#[derive(Parameters)]
#[parameters(num=T, device=D)]
pub struct SimpleModel<T: Num, D: Device> {
    pub linear_1: Linear<T, D>,
    pub linear_2: Linear<T, D>,
}

impl<D: Device> SimpleModel<f32, D> {
    #[must_use]
    pub fn new() -> Self {
        Self {
            linear_1: Linear::new(28 * 28, 512, true),
            linear_2: Linear::new(512, 10, true),
        }
    }
}

impl<D: Device> Default for SimpleModel<f32, D> {
    fn default() -> Self {
        Self::new()
    }
}

impl<D: Device> Module<f32, D> for SimpleModel<f32, D> {
    type Input = Variable<f32, D>;
    type Output = Variable<f32, D>;

    fn call(&self, inputs: Variable<f32, D>) -> Variable<f32, D> {
        let x = self.linear_1.call(inputs);
        let x = relu(x);
        self.linear_2.call(x)
    }
}

struct MnistDataset {
    data: Vec<(Vec<u8>, u8)>,
}

impl MnistDataset {
    fn new(input: &str, ans: &str) -> Self {
        // inputはcsvで一行目はヘッダーなので無視
        let input = std::fs::read_to_string(input).unwrap();
        let ans = std::fs::read_to_string(ans).unwrap();
        let mut data = Vec::new();
        for (_, (input, ans)) in input.lines().skip(1).zip(ans.lines()).enumerate() {
            let input = input
                .split(',')
                .map(|x| x.parse::<u8>().unwrap())
                .collect::<Vec<_>>();
            let ans = ans.parse::<u8>().unwrap();
            data.push((input, ans));
        }
        Self { data }
    }
}

impl Dataset<f32> for MnistDataset {
    type Item = (Vec<u8>, u8);

    fn item(&self, item: usize) -> Vec<Variable<f32, Cpu>> {
        let (x, y) = &self.data[item];
        let x_f32 = x.iter().map(|&x| f32::from(x)).collect::<Vec<_>>();
        let x = from_vec::<f32, _, Cpu>(x_f32, [784]);
        x.get_data_mut().to_ref_mut().div_scalar_assign(127.5);
        x.get_data_mut().to_ref_mut().sub_scalar_assign(1.0);
        let y_onehot = (0..10)
            .map(|i| if i == *y as usize { 1.0 } else { 0.0 })
            .collect::<Vec<_>>();
        let y = from_vec(y_onehot, [10]);
        vec![x, y]
    }

    fn len(&self) -> usize {
        self.data.len()
    }

    fn all_data(&mut self) -> &mut [Self::Item] {
        &mut self.data as &mut [Self::Item]
    }
}

#[expect(clippy::cast_precision_loss)]
fn main() {
    let model = SimpleModel::<f32, Nvidia>::new();
    let train = MnistDataset::new(
        "./zenu/examples/mnist_train_flattened.txt",
        "./zenu/examples/mnist_train_labels.txt",
    );
    let test = MnistDataset::new(
        "./zenu/examples/mnist_test_flattened.txt",
        "./zenu/examples/mnist_test_labels.txt",
    );
    let (train, val) = train_val_split(&train.data, 0.8, true);

    let test_dataloader = DataLoader::new(test, 1);

    let optimizer = SGD::<f32, Nvidia>::new(0.001);

    for num_epoch in 0..20 {
        println!("Epoch: {num_epoch}");
        set_train();
        let mut train_dataloader = DataLoader::new(
            MnistDataset {
                data: train.clone(),
            },
            1024,
        );

        train_dataloader.shuffle();

        let mut train_loss = 0.0;
        let mut num_iter = 0;

        for batch in train_dataloader {
            let input = batch[0].clone();
            let target = batch[1].clone();
            let input = input.to();
            let pred = model.call(input);
            let target = target.to();
            let loss = cross_entropy(pred, target);
            loss.backward();
            optimizer.update(&model);
            let loss_asum = loss.get_data().asum();
            loss.clear_grad();
            train_loss += loss_asum;
            num_iter += 1;
        }
        train_loss /= num_iter as f32;

        let val_loader = DataLoader::new(MnistDataset { data: val.clone() }, 1);

        no_train();
        let mut val_loss = 0.0;
        let mut num_iter = 0;
        for batch in val_loader {
            let input = batch[0].clone();
            let target = batch[1].clone();
            let pred = model.call(input.to());
            let loss = cross_entropy(pred.clone(), target.clone().to());
            val_loss += loss.get_data().asum();
            num_iter += 1;
        }
        val_loss /= num_iter as f32;

        println!("Epoch: {num_epoch}, Train Loss: {train_loss}, Val Loss: {val_loss}");
    }

    let mut test_loss = 0.;
    let mut num_iter = 0;

    for batch in test_dataloader {
        let input = batch[0].clone();
        let target = batch[1].clone();
        let pred = model.call(input.to());
        let loss = cross_entropy(pred, target.to());
        test_loss += loss.get_data().asum();
        num_iter += 1;
    }

    println!("Test Loss: {}", test_loss / num_iter as f32);
}

File: zenu/examples/mnist.py
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.data import Dataset, DataLoader, random_split
import csv


class MnistCSVDataset(Dataset):
    """
    mnist_train_flattened.txt などの入力画像データと
    mnist_train_labels.txt などのラベルデータを読み込む Dataset 実装例です。
    """
    def __init__(self, data_path: str, label_path: str):
        self.data = []
        # 画像データの読み込み（1行目がヘッダなのでスキップ）
        with open(data_path, "r", newline="") as f_data:
            reader_data = csv.reader(f_data)
            next(reader_data)  # ヘッダをスキップ
            data_lines = list(reader_data)

        # ラベルデータの読み込み（こちらはヘッダの有無に応じて適宜調整）
        with open(label_path, "r", newline="") as f_label:
            reader_label = csv.reader(f_label)
            label_lines = list(reader_label)

        # 行数が同じ前提で、(ピクセル配列, ラベル) をまとめて格納
        for line_data, line_label in zip(data_lines, label_lines):
            # ピクセル値を 0〜255 の u8 相当から float に変換
            # Rust の実装例では (値 / 127.5) - 1.0 で -1〜1 に正規化
            pixels = list(map(float, line_data))  # 文字列 → float
            label = int(line_label[0])           # ラベル（0〜9）
            self.data.append((pixels, label))

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx):
        x, y = self.data[idx]
        # Tensor に変換 & 正規化
        x = torch.tensor(x, dtype=torch.float32) / 127.5 - 1.0
        # CrossEntropyLoss を使う場合、ラベルは one-hot ではなく整数カテゴリにしておく
        y = torch.tensor(y, dtype=torch.long)
        return x, y


class SimpleModel(nn.Module):
    """
    Rust で定義していた SimpleModel<f32, D> 相当。
    28×28=784 次元→512 次元→10 次元の MLP。
    """
    def __init__(self):
        super().__init__()
        self.linear_1 = nn.Linear(784, 512)
        self.linear_2 = nn.Linear(512, 10)

    def forward(self, x):
        # x は [N, 784] 形状想定
        x = self.linear_1(x)
        x = F.relu(x)
        x = self.linear_2(x)
        return x


def main():
    # Dataset の作成
    train_data_path = "./zenu/examples/mnist_train_flattened.txt"
    train_label_path = "./zenu/examples/mnist_train_labels.txt"
    test_data_path = "./zenu/examples/mnist_test_flattened.txt"
    test_label_path = "./zenu/examples/mnist_test_labels.txt"

    # 1) 訓練＋バリデーション用データセット
    full_dataset = MnistCSVDataset(train_data_path, train_label_path)
    # 2) テストデータセット
    test_dataset = MnistCSVDataset(test_data_path, test_label_path)

    # train : val = 8 : 2 の割合で分割（Rust コードの train_val_split と同様）
    train_size = int(0.8 * len(full_dataset))
    val_size = len(full_dataset) - train_size
    train_dataset, val_dataset = random_split(full_dataset, [train_size, val_size])

    # DataLoader の作成
    train_loader = DataLoader(train_dataset, batch_size=32, shuffle=True)
    val_loader = DataLoader(val_dataset, batch_size=1, shuffle=False)
    test_loader = DataLoader(test_dataset, batch_size=1, shuffle=False)

    # モデル、損失関数、オプティマイザの準備
    model = SimpleModel()
    criterion = nn.CrossEntropyLoss()
    optimizer = torch.optim.SGD(model.parameters(), lr=0.001)

    # 学習ループ
    num_epochs = 20
    for epoch in range(num_epochs):
        # -------------------
        # Train
        # -------------------
        model.train()
        train_loss = 0.0
        for batch in train_loader:
            inputs, targets = batch
            # 順伝播
            outputs = model(inputs)
            loss = criterion(outputs, targets)
            # 逆伝播
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()
            # 損失の集計
            train_loss += loss.item()
        train_loss /= len(train_loader)

        # -------------------
        # Validation
        # -------------------
        model.eval()
        val_loss = 0.0
        with torch.no_grad():
            for batch in val_loader:
                inputs, targets = batch
                outputs = model(inputs)
                loss = criterion(outputs, targets)
                val_loss += loss.item()
        val_loss /= len(val_loader)

        print(f"Epoch: {epoch}, Train Loss: {train_loss:.4f}, Val Loss: {val_loss:.4f}")

    # -------------------
    # Test
    # -------------------
    model.eval()
    test_loss = 0.0
    with torch.no_grad():
        for batch in test_loader:
            inputs, targets = batch
            outputs = model(inputs)
            loss = criterion(outputs, targets)
            test_loss += loss.item()
    test_loss /= len(test_loader)

    print(f"Test Loss: {test_loss:.4f}")


if __name__ == "__main__":
    main()

